cmd/cnab-run/status.go

```diff
@@ -1,27 +1,205 @@
 package main
 
 import (
+	"encoding/json"
+	"fmt"
+	"io"
 	"os"
+	"sort"
+	"strings"
+	"text/tabwriter"
 
 	"github.com/docker/app/internal"
+	"github.com/docker/cli/cli/command"
 	"github.com/docker/cli/cli/command/stack"
 	"github.com/docker/cli/cli/command/stack/options"
 	"github.com/docker/cli/opts"
+	"github.com/docker/distribution/reference"
+	swarmtypes "github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/pkg/stringid"
 	"github.com/pkg/errors"
 )
 
+var (
+	listColumns = []struct {
+		header string
+		value  func(s *swarmtypes.Service) string
+	}{
+		{"ID", func(s *swarmtypes.Service) string { return stringid.TruncateID(s.ID) }},
+		{"NAME", func(s *swarmtypes.Service) string { return s.Spec.Name }},
+		{"MODE", func(s *swarmtypes.Service) string {
+			if s.Spec.Mode.Replicated != nil {
+				return "replicated"
+			}
+			return "global"
+		}},
+		{"REPLICAS", func(s *swarmtypes.Service) string {
+			if s.Spec.Mode.Replicated != nil {
+				return fmt.Sprintf("%d/%d", s.ServiceStatus.RunningTasks, s.ServiceStatus.DesiredTasks)
+			}
+			return ""
+		}},
+		{"IMAGE", func(s *swarmtypes.Service) string {
+			ref, err := reference.ParseAnyReference(s.Spec.TaskTemplate.ContainerSpec.Image)
+			if err != nil {
+				return "N/A"
+			}
+			if namedRef, ok := ref.(reference.Named); ok {
+				return reference.FamiliarName(namedRef)
+			}
+			return reference.FamiliarString(ref)
+		}},
+		{"PORTS", func(s *swarmtypes.Service) string {
+			return Ports(s.Endpoint.Ports)
+		}},
+	}
+)
+
+type portRange struct {
+	pStart   uint32
+	pEnd     uint32
+	tStart   uint32
+	tEnd     uint32
+	protocol swarmtypes.PortConfigProtocol
+}
+
+func (pr portRange) String() string {
+	var (
+		pub string
+		tgt string
+	)
+
+	if pr.pEnd > pr.pStart {
+		pub = fmt.Sprintf("%d-%d", pr.pStart, pr.pEnd)
+	} else {
+		pub = fmt.Sprintf("%d", pr.pStart)
+	}
+	if pr.tEnd > pr.tStart {
+		tgt = fmt.Sprintf("%d-%d", pr.tStart, pr.tEnd)
+	} else {
+		tgt = fmt.Sprintf("%d", pr.tStart)
+	}
+	return fmt.Sprintf("*:%s->%s/%s", pub, tgt, pr.protocol)
+}
+
+// Ports formats port configuration. This function is copied et adapted from docker CLI
+// see https://github.com/docker/cli/blob/d6edc912ce/cli/command/service/formatter.go#L655
+func Ports(servicePorts []swarmtypes.PortConfig) string {
```

Comment:
There is really no need for this function to be public

Reply:
You're right. Fixed